### PR TITLE
Feature tests for minimum APIs needed by user's browser

### PR DESCRIPTION
**Pull request by closuresandcoffee**
#### Changes
I created a utility module that has methods which can be used to check for browser support of the APIs in question. Please note that I just did checks for window level objects and methods, and didn't go too far into the inner workings of those APIs to make sure I could instantiate those interfaces or anything. Let me know if I need to dive deeper and I can.

I also tweaked the notification `show` method. I needed to make a version of the modal without buttons and the ability to add another parameter and the arguments were getting pretty crowded. I hope you don't mind! I can always find another approach if need be, I just felt this was the cleanest and since it is only used in a couple of spots I felt like it was a good improvement to make now versus later. 😅 

Fixes: https://github.com/getify/youperiod.app/issues/58

#### Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [X] This change requires a documentation update

I was on the fence on whether or not to update `SPECS.md` since `browser-support.js` is really only going to be used for this one piece.

#### Checklist

- [X] My code follows the style guidelines of this project
- [X] I have performed a self-review of my own code
- [X] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [X] My changes generate no new warnings

### CODE REVIEW

File: web/js/browser-support.js [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+	return "Worker" in window;
+}
+
+function indexedDbSupported() {
+	return "indexedDB" in window
+		&& typeof window.indexedDB.open === "function";
+}
+
+function sessionStorageSupported() {
+	var localStorageAvailable = "localStorage" in window
```
Comment (getify) [inline web/js/browser-support.js:20]: AFAIK, no browsers have one and not the other, but... in our case, we only ever plan to use `sessionStorage`, so I think we can safely skip the test for `localStorage`.

File: web/js/notification-manager.js [hunk 2]
```diff
@@ -17,16 +18,30 @@ function init(mainDOMElement) {
...
 	cancelBtnEl.addEventListener("click",() => hide(/*result=*/false),false);
 }
 
-function show(msg,isModal = false,isError = false,showCancel = false,onClose = () => {}) {
-	closeCallback = onClose;
+/**
+ * Displays a notification to the end-user
+ * @param {string} message
+ * @param {Object} options
+ * @param {boolean} options.isModal if isModal is true, clear after a set duration
```
Comment (getify) [inline web/js/notification-manager.js:33]: Perhaps I'm misunderstanding, but "clear after a set duration" is when the notification is NOT modal, I think. In this implementation, I think "modal" means "must click a button to dismiss".

File: web/js/browser-support.js [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+
+function optionalFeaturesSupported() {
+	return (
+		notificationApiSupported()
+		&& credentialsContainerSupported()
+		&& getUserMediaSupported()
+	);
+}
+
+export { minimumFeaturesSupported, optionalFeaturesSupported };
```
Comment (getify) [inline web/js/browser-support.js:78]: Should this module just auto-run these tests, and export the `true` / `false` results? I'm not seeing any reason to defer and require the including code to have to invoke the functions.

File: web/js/notification-manager.js [hunk 4]
```diff
@@ -17,16 +18,30 @@ function init(mainDOMElement) {
...
 	cancelBtnEl = bannerEl.querySelector("button.cancel-btn");
 	okBtnEl = bannerEl.querySelector("button[type=submit]");
 
 	formEl.addEventListener("submit",onFormSubmit,false);
 	cancelBtnEl.addEventListener("click",() => hide(/*result=*/false),false);
 }
 
-function show(msg,isModal = false,isError = false,showCancel = false,onClose = () => {}) {
-	closeCallback = onClose;
+/**
```
Comment (getify) [inline web/js/notification-manager.js:29]: I am glad you made this change from the individual params to the options object. I had intended to do so myself! :)

File: web/js/notification-manager.js [hunk 5]
```diff
@@ -76,6 +91,48 @@ function show(msg,isModal = false,isError = false,showCancel = false,onClose = (
 	else {
 		cancelBtnEl.classList.add("hidden");
 	}
+
+	if (canDismiss) {
+		addEventListeners(isModal, showCancel);
+	}
+	else {
+		okBtnEl.classList.add("hidden");
+		removeEventListeners();
```
Comment (getify) [inline web/js/notification-manager.js:100]: Is the call to `removeEventListeners()` here needed? I think `hide()` always unbinds them, and I think `hide()` is called above if `show(..)` has been invoked while the notification is already visible.

File: web/js/notification-manager.js [hunk 6]
```diff
@@ -76,6 +91,48 @@ function show(msg,isModal = false,isError = false,showCancel = false,onClose = (
...
+	}
+}
+
+function addEventListeners(isModal, showCancel) {
```
Comment (getify) [inline web/js/notification-manager.js:104]: Good refactoring to move this logic to its own function, thanks.

**Review by getify — CHANGES_REQUESTED** [review]
Looks pretty good, thanks. Few minor things to address.

File: web/js/notification-manager.js (hunk 2 above)
Comment (closuresandcoffee) [reply web/js/notification-manager.js:33]: You are correct. I'll get that updated.

File: web/js/browser-support.js (hunk 3 above)
Comment (closuresandcoffee) [reply web/js/browser-support.js:78]: I don't have a preference either way. If you want me to change that, I can, just let me know!

File: web/js/notification-manager.js (hunk 4 above)
Comment (closuresandcoffee) [reply web/js/notification-manager.js:29]: Less params the better! Glad you thought it was a good choice.

File: web/js/browser-support.js (hunk 3 above)
Comment (getify) [reply web/js/browser-support.js:78]: I think for now, it's the simpler option. We can always change it back later if we need more sophistication.

File: web/js/notification-manager.js (hunk 5 above)
Comment (closuresandcoffee) [reply web/js/notification-manager.js:100]: I will remove that method and the related call.

**Comment by closuresandcoffee** [comment]
@getify Changes have been pushed!

File: web/js/browser-support.js [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+function webCryptoSupported() {
+	return "crypto" in window
+		&& typeof window.crypto.randomUUID === "function";
+}
+
+function textEncodeDecodeSupported() {
+	return "TextEncoder" in window && "TextDecoder" in window;
+}
+
+function storeManagerPersistSupported() {
```
Comment (getify) [inline web/js/browser-support.js:33]:
```suggestion
function storageManagerPersistSupported() {
```

File: web/js/browser-support.js [hunk 8]
```diff
@@ -0,0 +1,69 @@
...
+	return "credentials" in window.navigator
+		&& typeof window.navigator.credentials.get === "function";
+}
+
+function getUserMediaSupported() {
+	return "mediaDevices" in window.navigator
+		&& typeof window.navigator.mediaDevices.getUserMedia === "function";
+}
+
+var MINIMUM_FEATURES_SUPPORTED = (
```
Comment (getify) [inline web/js/browser-support.js:52]:
```suggestion
cpnst MINIMUM_FEATURES_SUPPORTED = (
```

File: web/js/browser-support.js [hunk 9]
```diff
@@ -0,0 +1,69 @@
...
+	webAssemblySupported()
+	&& serviceWorkersSupported()
+	&& webWorkersSupported()
+	&& indexedDbSupported()
+	&& sessionStorageSupported()
+	&& webCryptoSupported()
+	&& textEncodeDecodeSupported()
+	&& storeManagerPersistSupported()
```
Comment (getify) [inline web/js/browser-support.js:60]:
```suggestion
	&& storageManagerPersistSupported()
```

**Review by getify — APPROVED** [review]
LGTM, thanks!

**Pull request merged**